pkg/trie/triedb/triedb.go

```diff
@@ -114,4 +175,290 @@ func (t *TrieDB) getNode(
+		action, err := inspector(n.node, key)
+		if err != nil {
+			return nil, false, err
+		}
+		switch a := action.(type) {
+		case Restore:
+			return Cached{a.node, n.hash}, false, nil
+		case Replace:
+			t.deathRow[n.hash] = nil
+			return NewNewNode(a.node), true, nil
+		case Delete:
+			t.deathRow[n.hash] = nil
+			return nil, false, nil
+		default:
+			panic("unreachable")
+		}
+	default:
+		panic("unreachable")
+	}
+}
+
+// insertInspector inserts the new key / value pair into the given node `stored`
+func (t *TrieDB) insertInspector(stored Node, keyNibbles []byte, value []byte, oldValue *Value) (Action, error) {
+	partial := keyNibbles
+
+	switch n := stored.(type) {
+	case Empty:
+		// If the node is empty we have to replace it with a leaf node with the
+		// new value
+		value := NewValue(value, t.layout.MaxInlineValue())
+		return Replace{node: Leaf{partialKey: partial, value: value}}, nil
+	case Leaf:
+		existingKey := n.partialKey
+		common := nibbles.CommonPrefix(partial, existingKey)
+
+		if common == len(existingKey) && common == len(partial) {
+			// We are trying to insert a value in the same leaf so we just need
+			// to replace the value
+			value := NewValue(value, t.layout.MaxInlineValue())
+			unchanged := n.value == value
+			t.replaceOldValue(oldValue, n.value)
+			leaf := Leaf{partialKey: n.partialKey, value: n.value}
+			if unchanged {
+				// If the value didn't change we can restore this leaf previously
+				// taken from storage
+				return Restore{leaf}, nil
+			}
+			return Replace{leaf}, nil
+		} else if common < len(existingKey) {
+			// If the common prefix is less than this leaf's key then we need to
+			// create a branch node. Then add this leaf and the new value to the
+			// branch
+			var children [codec.ChildrenCapacity]NodeHandle
+
+			idx := existingKey[common]
+
+			// Modify the existing leaf partial key and add it as a child
+			newLeaf := Leaf{existingKey[common+1:], n.value}
+			children[idx] = t.storage.alloc(New{node: newLeaf}).toNodeHandle()
+			branch := Branch{
+				partialKey: partial[:common],
+				children:   children,
+				value:      nil,
+			}
+
+			// Use the inspector to add the new leaf as part of this branch
+			branchAction, err := t.insertInspector(branch, keyNibbles, value, oldValue)
+			if err != nil {
+				return nil, err
+			}
+			return Replace{branchAction.getNode()}, nil
+		} else {
+			// we have a common prefix but the new key is longer than the existing
+			// then we turn this leaf into a branch and add the new leaf as a child
+			var branch Node = Branch{
+				partialKey: n.partialKey,
+				children:   [codec.ChildrenCapacity]NodeHandle{},
+				value:      n.value,
+			}
+			// Use the inspector to add the new leaf as part of this branch
+			// And replace the node with the new branch
+			action, err := t.insertInspector(branch, keyNibbles, value, oldValue)
+			if err != nil {
+				return nil, err
+			}
+			branch = action.getNode()
+			return Replace{branch}, nil
```

Comment:
Is Deletion just the same as a replacement with nil? This action could be a delete right? In which case we get `Replace{nil}` and I wanted to make sure that was intended

Reply:
mmm it's not actually the same, but I'll address deletions in my next PR